# <span style="color: red;">A Statistical Analysis of the FIFA World Cup 2022 Player Data</span>

Arguably one of the greatest World Cups in history. Together, we will take a deep dive into the achievements of the individual players and teams throughout this electrifying campaign.

# <span style="color: red;">Project Goals and Project Data Questions</span>
__Formatting+Cleaning:__
- To reassign the names of columns
- To drop unwanted columns
- Reformat the age to be in days
- To add any new columns if necessary
- To create a new DataFrame if necessary
    
__We're going to find out...__
- The highest scorer
- The highest assist maker
- The player with the most goals and assists
- The player with the most goals and assists per 90 minutes
- The player with the most minutes played
- The youngest and oldest players
- The youngest and oldest players to score
- The most red carded player
- What country scored the most goals

__Data Sources:__
<br>*player_stats.csv*
<br>https://www.kaggle.com/datasets/swaptr/fifa-world-cup-2022-player-data/data
<br>*fifa_wc_2022_player_stats.csv*
<br>https://www.kaggle.com/datasets/rhugvedbhojane/fifa-world-cup-2022-players-statistics/data

__Licenses:__
<br>*player_stats.csv*
<br>https://opendatacommons.org/licenses/odbl/1-0/
<br>*fifa_wc_2022_player_stats.csv*
<br>https://creativecommons.org/licenses/by-nc-sa/4.0/

# <span style="color: red;">Import and Clean Data</span>

### <span style="color: red;">Imported libraries</span>

In [13359]:
import pandas as pd
from unidecode import unidecode

### <span style="color: red;">Dataset Loading and Inspection</span>

The player and team data was stored in the following CSV files
- **player_stats.csv**
- **fifa_wc_2022_player_stats.csv**

Let's load the first dataset below.

In [13360]:
wc_data_one = pd.read_csv('/Users/admin/world_cup_2022_project/world_cup_2022_analysis/player_stats.csv')
print(wc_data_one.info())
wc_data_one.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 680 entries, 0 to 679
Data columns (total 31 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   player                    680 non-null    object 
 1   position                  680 non-null    object 
 2   team                      680 non-null    object 
 3   age                       680 non-null    object 
 4   club                      679 non-null    object 
 5   birth_year                680 non-null    int64  
 6   games                     680 non-null    int64  
 7   games_starts              680 non-null    int64  
 8   minutes                   680 non-null    int64  
 9   minutes_90s               680 non-null    float64
 10  goals                     680 non-null    int64  
 11  assists                   680 non-null    int64  
 12  goals_pens                680 non-null    int64  
 13  pens_made                 680 non-null    int64  
 14  pens_att  

player position       team     age        club  birth_year  \
0            Aaron Mooy       MF  Australia  32-094      Celtic        1990   
1          Aaron Ramsey       MF      Wales  31-357        Nice        1990   
2     Abdelhamid Sabiri       MF    Morocco  26-020   Sampdoria        1996   
3     Abdelkarim Hassan       DF      Qatar  29-112  Al Sadd SC        1993   
4  Abderrazak Hamdallah       FW    Morocco  32-001  Al-Ittihad        1990   

   games  games_starts  minutes  minutes_90s  ...  goals_assists_pens_per90  \
0      4             4      360          4.0  ...                       0.0   
1      3             3      266          3.0  ...                       0.0   
2      5             2      181          2.0  ...                       0.5   
3      3             3      270          3.0  ...                       0.0   
4      4             0       68          0.8  ...                       0.0   

    xg  npxg  xg_assist  npxg_xg_assist  xg_per90  xg_assist_per90  \
0  0.0   0.0        0.1             0.1      0.01             0.02   
1  0.0   0.0        0.0             0.1      0.01             0.01   
2  0.1   0.1        0.9             1.0      0.08             0.53   
3  0.3   0.3        0.0             0.3      0.10             0.01   
4  0.4   0.4        0.0             0.4      0.52             0.00   

   xg_xg_assist_per90  npxg_per90  npxg_xg_assist_per90  
0                0.03        0.01                  0.03  
1                0.02        0.01                  0.02  
2                0.60        0.08                  0.60  
3                0.11        0.10                  0.11  
4                0.52        0.52                  0.52  

[5 rows x 31 columns]

Initial observations:
- There are results for 680 players in the tournament
- There are missing values for several of the columns
- There are some potential formatting problems (age column)
- There are a lot of columns (31)

We can then load the second dataset for inspection.

In [13361]:
wc_data_two = pd.read_csv('/Users/admin/world_cup_2022_project/world_cup_2022_analysis/fifa_wc_2022_player_stats.csv')
print(wc_data_two.info())
wc_data_two.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 814 entries, 0 to 813
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Nationality                  814 non-null    object 
 1   FIFA Ranking                 814 non-null    int64  
 2   National Team Kit Sponsor    813 non-null    object 
 3   Position                     814 non-null    object 
 4   National Team Jersey Number  813 non-null    float64
 5   Player DOB                   814 non-null    object 
 6   Club                         814 non-null    object 
 7   Player Name                  814 non-null    object 
 8    Appearances                 814 non-null    object 
 9   Goals Scored                 813 non-null    object 
 10  Assists Provided             813 non-null    object 
 11  Dribbles per 90              813 non-null    object 
 12  Interceptions per 90         813 non-null    object 
 13  Tackles per 90      

Nationality   FIFA Ranking  National Team Kit Sponsor Position  \
0    Argentina              2                    Adidas       GK   
1    Argentina              2                    Adidas       GK   
2    Argentina              2                    Adidas       GK   
3    Argentina              2                    Adidas       DF   
4    Argentina              2                    Adidas       DF   

   National Team Jersey Number    Player DOB        Club        Player Name   \
0                         23.0   Sep 2, 1992  Aston Villa  Emiliano Martinez   
1                          1.0  Oct 16, 1986        River      Franco Armani   
2                         12.0  May 20, 1992   Villarreal     Geronimo Rulli   
3                         19.0  Feb 12, 1988      Benfica   Nicolas Otamendi   
4                          8.0  Oct 28, 1991      Sevilla       Marcos Acuna   

   Appearances Goals Scored  Assists Provided  Dribbles per 90  \
0            7             0                 0            0.00   
1            0             -                 -               -   
2            0             -                 -               -   
3            7             0                 1            0.33   
4            6             0                 0            1.45   

  Interceptions per 90 Tackles per 90 Total Duels Won per 90 Save Percentage  \
0                 0.00           0.00                   0.65          46.67%   
1                    -              -                      -               -   
2                    -              -                      -               -   
3                 1.17           1.30                   7.17               -   
4                 0.48           2.90                   7.97               -   

  Clean Sheets Brand Sponsor/Brand Used  
0          43%                   Adidas  
1            -                     Nike  
2            -                   Adidas  
3            -                     Nike  
4            -                     Nike

Initial observations:
- We have data from 814 players in the tournament
- There is a null value in the goals scored column
- Some of the values in the goals scored column are dashes
- There are some potential formatting problems (trailing whitespace and capitilization)

### <span style="color: red;">Data Cleaning and Preparation</span>

To start, we can obtain a list of the column names from each DataFrame, and names the values using from this list that we want to keep for analysis.

In [13362]:
wc_data_one.columns

Index(['player', 'position', 'team', 'age', 'club', 'birth_year', 'games',
       'games_starts', 'minutes', 'minutes_90s', 'goals', 'assists',
       'goals_pens', 'pens_made', 'pens_att', 'cards_yellow', 'cards_red',
       'goals_per90', 'assists_per90', 'goals_assists_per90',
       'goals_pens_per90', 'goals_assists_pens_per90', 'xg', 'npxg',
       'xg_assist', 'npxg_xg_assist', 'xg_per90', 'xg_assist_per90',
       'xg_xg_assist_per90', 'npxg_per90', 'npxg_xg_assist_per90'],
      dtype='object')

In [13363]:
# hand picked values from stats_list to keep for the DataFrame, then used .drop to eliminate unwanted values

dropped_columns_one = ['games', 'birth_year', 'position', 'games_starts', 'minutes_90s','goals_pens', 'pens_made', 'pens_att',
                      'goals_pens_per90', 'goals_assists_pens_per90', 'xg', 'npxg', 'xg_assist','npxg_xg_assist', 'cards_yellow',
                      'xg_per90', 'xg_assist_per90', 'xg_xg_assist_per90', 'npxg_per90', 'npxg_xg_assist_per90', 
                      'goals_per90', 'assists_per90', 'goals_assists_per90']

# kept 'player', 'position', 'team', 'age', 'club', 'minutes', 'goals', 'assists', 'cards_red'

wc_data_one = wc_data_one.drop(labels=dropped_columns_one, axis=1)

In [13364]:
wc_data_two.columns

Index(['Nationality ', 'FIFA Ranking ', 'National Team Kit Sponsor',
       'Position', 'National Team Jersey Number', 'Player DOB', 'Club ',
       'Player Name ', ' Appearances', 'Goals Scored ', 'Assists Provided ',
       'Dribbles per 90', 'Interceptions per 90', 'Tackles per 90',
       'Total Duels Won per 90', 'Save Percentage', 'Clean Sheets',
       'Brand Sponsor/Brand Used'],
      dtype='object')

In [13365]:
dropped_columns_two = ['FIFA Ranking ', 'National Team Kit Sponsor',
       'National Team Jersey Number', 'Player DOB',
       ' Appearances', 'Assists Provided ', 'Dribbles per 90', 
       'Interceptions per 90', 'Tackles per 90', 'Total Duels Won per 90', 
       'Save Percentage', 'Clean Sheets', 'Brand Sponsor/Brand Used']

# kept 'Nationality ', 'Player Name ', 'Club ', 'Goals Scored ' and 'Position'

wc_data_two = wc_data_two.drop(labels=dropped_columns_two, axis=1)

We can then rename any columns that contain special characters/don't match the original DataFrame.

In [13366]:
# Renamed columns
renamed_columns_one = {'cards_red': 'red cards'}

wc_data_one = wc_data_one.rename(mapper=renamed_columns_one, axis=1)

wc_data_one.head()

player       team     age        club  minutes  goals  \
0            Aaron Mooy  Australia  32-094      Celtic      360      0   
1          Aaron Ramsey      Wales  31-357        Nice      266      0   
2     Abdelhamid Sabiri    Morocco  26-020   Sampdoria      181      0   
3     Abdelkarim Hassan      Qatar  29-112  Al Sadd SC      270      0   
4  Abderrazak Hamdallah    Morocco  32-001  Al-Ittihad       68      0   

   assists  red cards  
0        0          0  
1        0          0  
2        1          0  
3        0          0  
4        0          0

In [13367]:
renamed_columns_two = {'Nationality ': 'team',	
                       'Player Name ': 'player',
                       'Goals Scored ': 'goals',
                       'Club ': 'club',
                       'Position' : 'position'}

wc_data_two = wc_data_two.rename(mapper=renamed_columns_two, 
                         axis=1)

wc_data_two.head()

team position         club             player goals
0  Argentina       GK  Aston Villa  Emiliano Martinez     0
1  Argentina       GK        River      Franco Armani     -
2  Argentina       GK   Villarreal     Geronimo Rulli     -
3  Argentina       DF      Benfica   Nicolas Otamendi     0
4  Argentina       DF      Sevilla       Marcos Acuna     0

We can then clean the data further by replacing any string or null values in the goals column with zero

In [13368]:
is_dash = (wc_data_two['goals'] == '-')
wc_data_two[is_dash].head()

team position             club           player goals
1   Argentina       GK            River    Franco Armani     -
2   Argentina       GK       Villarreal   Geronimo Rulli     -
26     France       GK        Tottenham      Hugo Lloris     -
27     France       GK  West Ham United  Alphonse Areola     -
28     France       GK           Rennes   Steve Mandanda     -

In [13369]:
# replaced '-' with 0

wc_data_two['goals'] = wc_data_two['goals'].replace('-', '0')
wc_data_two[is_dash].head()

team position             club           player goals
1   Argentina       GK            River    Franco Armani     0
2   Argentina       GK       Villarreal   Geronimo Rulli     0
26     France       GK        Tottenham      Hugo Lloris     0
27     France       GK  West Ham United  Alphonse Areola     0
28     France       GK           Rennes   Steve Mandanda     0

In [13370]:
is_NA_one = (wc_data_two['goals'] == 'N.A')
wc_data_two[is_NA_one].head()

team position                 club           player goals
54        Croatia       DF           AEK Athens     Domagoj Vida   N.A
61        Croatia       DF             Sassuolo     Martin Erlic   N.A
69        Croatia       MF  Eintracht Frankfurt  Kristijan Jakic   N.A
70        Croatia       MF    Red Bull Salzburg       Luka Sucic   N.A
107  Netherlands        DF                Inter   Stefan de Vrij   N.A

In [13371]:
is_NA_two = (wc_data_two['goals'] == 'N,A')
wc_data_two[is_NA_two]

team position           club        player goals
340  South Korea       DF  Ulsan Hyundai  Kim Tae-hwan   N,A

There were a further 147 values in the goals column that were NaN, 'N.A' and 'N,A', so we can use the same methods to replace them with 0.

In [13372]:
wc_data_two['goals'] = wc_data_two['goals'].replace('N.A', '0')

wc_data_two['goals'] = wc_data_two['goals'].replace('N,A', '0')

Let's verify the change through displaying the goals columns below:

In [13373]:
wc_data_two[is_NA_one].head()

team position                 club           player goals
54        Croatia       DF           AEK Athens     Domagoj Vida     0
61        Croatia       DF             Sassuolo     Martin Erlic     0
69        Croatia       MF  Eintracht Frankfurt  Kristijan Jakic     0
70        Croatia       MF    Red Bull Salzburg       Luka Sucic     0
107  Netherlands        DF                Inter   Stefan de Vrij     0

In [13374]:
wc_data_two[is_NA_two].head()

team position           club        player goals
340  South Korea       DF  Ulsan Hyundai  Kim Tae-hwan     0

The null value was still in the DataFrame somewhere. We can use .isnull to identify the row where this occurs.

In [13375]:
null_values = wc_data_two['goals'].isnull()
wc_data_two[null_values]

team position    club                  player goals
713  Serbia       GK  Torino  Vanja Milinkovic-Savic   NaN

We can change it by selecting a specific row using pandas' .loc method.

In [13376]:
# found here that the null value belonged to Vanja Milinkovic-Savic, and so changed it

wc_data_two.loc[[713],['goals']] = '0'

Let's verify the change.

In [13377]:
wc_data_two[null_values]

team position    club                  player goals
713  Serbia       GK  Torino  Vanja Milinkovic-Savic     0

Using .isnull again, we can take a deeper look at the null value from the club column.

In [13378]:
null_club = wc_data_one[wc_data_one['club'].isnull()]
null_club

player     team     age club  minutes  goals  assists  red cards
354  Kevin Rodríguez  Ecuador  22-289  NaN        2      0        0          0

Kevin Rodríguez, who do you play for? Let's ask google:
<br>After a quick look at his wiki, we find out that at the time of the World Cup, he played for *Imbabura SC*
<br>I implemented this into my DataFrame below:

__Source__
<br>https://es.wikipedia.org/wiki/Kevin_Rodr%C3%ADguez

We can change the specific row using pandas' .iloc method.


In [13379]:
wc_data_one.iloc[354, 3] = 'Imbabura SC'

# verified the change with .iloc

wc_data_one.iloc[354, 3]

'Imbabura SC'

We may want to modify some of the text data for tidiness and consistency - we can start by removing unnecessary dashes from clubs using .replace.

In [13380]:
wc_data_one['club'] = wc_data_one['club'].str.replace('-', '')
wc_data_one.assign()

player        team     age             club  minutes  \
0              Aaron Mooy   Australia  32-094           Celtic      360   
1            Aaron Ramsey       Wales  31-357             Nice      266   
2       Abdelhamid Sabiri     Morocco  26-020        Sampdoria      181   
3       Abdelkarim Hassan       Qatar  29-112       Al Sadd SC      270   
4    Abderrazak Hamdallah     Morocco  32-001        AlIttihad       68   
..                    ...         ...     ...              ...      ...   
675        Ángel Di María   Argentina  34-307         Juventus      288   
676       Ángelo Preciado     Ecuador  24-303             Genk      264   
677          Éder Militão      Brazil  24-334      Real Madrid      347   
678          Óscar Duarte  Costa Rica  33-198          AlWehda      270   
679        İlkay Gündoğan     Germany  32-055  Manchester City      189   

     goals  assists  red cards  
0        0        0          0  
1        0        0          0  
2        0        1          0  
3        0        0          0  
4        0        0          0  
..     ...      ...        ...  
675      1        1          0  
676      0        1          0  
677      0        0          0  
678      0        0          0  
679      1        0          0  

[680 rows x 8 columns]

Let's add the values from 'years' and 'days' together to specify age.

In [13381]:
wc_data_one['age'] = wc_data_one['age'].str.replace('-', ' years, ') + ' days'

We can then remove unnecessary zeros from the days in the age column.

In [13382]:
age_metrics = wc_data_one['age'].str.split(
    pat=',', 
    expand=True)

# used the apply function to apply lambda (a small anonymous function) to multiple rows within the DataFrame
# using x.split, split the values into two parts (number of days and 'days')
# stripping them of leading zeros with .lstrip
# joined them back together with .join
age_metrics[1] = age_metrics[1].apply(lambda x: ' '.join(part.lstrip('0') for part in x.split(' ')))

wc_data_one['age'] = age_metrics[0].str.cat(age_metrics[1], sep=', ')
wc_data_one.head()

player       team                  age        club  minutes  \
0            Aaron Mooy  Australia   32 years,  94 days      Celtic      360   
1          Aaron Ramsey      Wales  31 years,  357 days        Nice      266   
2     Abdelhamid Sabiri    Morocco   26 years,  20 days   Sampdoria      181   
3     Abdelkarim Hassan      Qatar  29 years,  112 days  Al Sadd SC      270   
4  Abderrazak Hamdallah    Morocco    32 years,  1 days   AlIttihad       68   

   goals  assists  red cards  
0      0        0          0  
1      0        0          0  
2      0        1          0  
3      0        0          0  
4      0        0          0

We need to add a few columns to the DataFrame in order to answer several of the Project Data Questions

Specifically, to add the following columns:
- Assists per 90 minutes
- Goals and assists
- Goals and assists per 90 minutes
- Age in days
- Red cards per 90 minutes

Assists were tied between three players, so we need to make a new column for assists per 90 minutes.

In [13383]:
top_assists_per_90 = wc_data_one.sort_values(by='assists', ascending=False).head(5)

top_assists_per_90['assists per 90'] = (wc_data_one['assists'] / wc_data_one['minutes'] * 90).round(decimals=3)

Let's make a new column by adding the goals and assist column values together.

In [13384]:
wc_data_one['goals+assists'] = wc_data_one['goals'] + wc_data_one['assists']

Another new column - goals + assists per 90 minutes.

In [13385]:
wc_data_one['goals+assists per 90'] = ((wc_data_one['goals'] + wc_data_one['assists']) / wc_data_one['minutes'] * 90).round(decimals=3)

Let's calculate the age in days to allow youngest and oldest players to be identified.

Calculating the age was awkward as the 'age' column is currently being treated as a string. We can solve this by splitting the column values into two by their comma, creating a new column.

In [13386]:
age_metrics = wc_data_one['age'].str.split(
    pat=',', 
    expand=True)

# removed the letter values from the string, then converted them to integer form for calculation
age_metrics[0] = age_metrics[0].str.replace('years', '').astype('int')
age_metrics[1] = age_metrics[1].str.replace('days', '').astype('int')
age_metrics

# calculated the age in days by multiplying the years by days per year then adding the leftover days to the result
wc_data_one['age in days'] = (age_metrics[0] * 365) + age_metrics[1]
wc_data_one.head()

player       team                  age        club  minutes  \
0            Aaron Mooy  Australia   32 years,  94 days      Celtic      360   
1          Aaron Ramsey      Wales  31 years,  357 days        Nice      266   
2     Abdelhamid Sabiri    Morocco   26 years,  20 days   Sampdoria      181   
3     Abdelkarim Hassan      Qatar  29 years,  112 days  Al Sadd SC      270   
4  Abderrazak Hamdallah    Morocco    32 years,  1 days   AlIttihad       68   

   goals  assists  red cards  goals+assists  goals+assists per 90  age in days  
0      0        0          0              0                 0.000        11774  
1      0        0          0              0                 0.000        11672  
2      0        1          0              1                 0.497         9510  
3      0        0          0              0                 0.000        10697  
4      0        0          0              0                 0.000        11681

We also need to calculate red cards per 90 minutes as these results were tied between four players.

In [13387]:
wc_data_one['red cards per 90'] = ((wc_data_one['red cards']) / wc_data_one['minutes'] * 90).round(decimals=3)

Now that all of the columns have been added, let's remove any unnecessary ones.

As well as changing the player names, we need to change the team names so that they match.

In [13388]:
wc_data_two['goals'] = wc_data_two['goals'].astype(int)

In [13389]:
unique_values_one = wc_data_one['team'].unique()
unique_values_one

array(['Australia', 'Wales', 'Morocco', 'Qatar', 'Senegal', 'Ghana',
       'Saudi Arabia', 'IR Iran', 'France', 'Uruguay', 'Tunisia',
       'Ecuador', 'Spain', 'Serbia', 'Brazil', 'Denmark', 'Argentina',
       'Mexico', 'Canada', 'Belgium', 'Cameroon', 'Croatia',
       'Netherlands', 'Portugal', 'Costa Rica', 'United States',
       'Germany', 'Japan', 'Switzerland', 'Poland', 'England',
       'Korea Republic'], dtype=object)

In [13390]:
unique_values_two = wc_data_two['team'].unique()
unique_values_two

array(['Argentina', 'France', 'Croatia', 'Morocco ', 'Netherlands ',
       'England', 'Brazil', 'Portugal', 'Australia', 'USA', 'Poland ',
       'Senegal', 'Japan', 'South Korea', 'Spain', 'Switzerland',
       'Ecuador ', 'Qatar', 'Iran', 'Wales', 'Mexico ', 'Saudi Arabia',
       'Denmark', 'Tunisia', 'Germany', 'Costa Rica', 'Belgium',
       'Cameroon', 'Serbia', 'Ghana', 'Uruguay', 'Canada'], dtype=object)

In [13391]:
countries_one = ['Australia', 'Wales', 'Morocco', 'Qatar', 'Senegal', 'Ghana', 'Saudi Arabia',
 'IR Iran', 'France', 'Uruguay', 'Tunisia', 'Ecuador', 'Spain', 'Serbia',
 'Brazil', 'Denmark', 'Argentina', 'Mexico', 'Canada', 'Belgium', 'Cameroon',
 'Croatia', 'Netherlands', 'Portugal', 'Costa Rica', 'United States', 'Germany',
 'Japan', 'Switzerland', 'Poland', 'England', 'Korea Republic']

countries_two = ['argentina', 'france', 'croatia', 'morocco', 'netherlands', 'england',
 'brazil', 'portugal', 'australia', 'usa', 'poland', 'senegal', 'japan',
 'south korea', 'spain', 'switzerland', 'ecuador', 'qatar', 'iran', 'wales',
 'mexico', 'saudi arabia', 'denmark', 'tunisia', 'germany', 'costa rica',
 'belgium', 'cameroon', 'serbia', 'ghana', 'uruguay', 'canada']

#countries_one = sorted(countries_one)
print(countries_one, "\n")
#countries_two = sorted(countries)
print(countries_two)

['Australia', 'Wales', 'Morocco', 'Qatar', 'Senegal', 'Ghana', 'Saudi Arabia', 'IR Iran', 'France', 'Uruguay', 'Tunisia', 'Ecuador', 'Spain', 'Serbia', 'Brazil', 'Denmark', 'Argentina', 'Mexico', 'Canada', 'Belgium', 'Cameroon', 'Croatia', 'Netherlands', 'Portugal', 'Costa Rica', 'United States', 'Germany', 'Japan', 'Switzerland', 'Poland', 'England', 'Korea Republic'] 

['argentina', 'france', 'croatia', 'morocco', 'netherlands', 'england', 'brazil', 'portugal', 'australia', 'usa', 'poland', 'senegal', 'japan', 'south korea', 'spain', 'switzerland', 'ecuador', 'qatar', 'iran', 'wales', 'mexico', 'saudi arabia', 'denmark', 'tunisia', 'germany', 'costa rica', 'belgium', 'cameroon', 'serbia', 'ghana', 'uruguay', 'canada']


To further clean the teams, we can iterate through each of them and check for any mismatched values. We can do this using a nested loop:

In [13392]:
# Iterate through the lists simultaneously
for country1 in countries_two:
    matched = False
    for country2 in countries_one:
        # Check if one country name is partially contained in the other
        if country1.lower() in country2.lower() and country2.lower() in country1.lower():
            matched = True
            break
    if not matched:
        print(f"Mismatch: {country1}")

Mismatch: usa
Mismatch: south korea
Mismatch: iran


It can be seen from this iteration that 'IR Iran', 'United States' and 'Korea Republic' didn't match the names of the original list ('iran', 'usa' and 'south korea'). We can change these values using pandas methods.

<br>In order to replace all these column values with one block of code, we can chain together multiple 'str.replace' calls.

In [13393]:
wc_data_one['team'] = wc_data_one['team'].str.replace('IR Iran', 'iran', regex=False)\
    .replace('United States', 'usa', regex=False)\
    .replace('Korea Republic', 'south korea', regex=False)

We obtain each country by calling .unique() on the team column.

In [13394]:
wc_data_two['team'].unique()

array(['Argentina', 'France', 'Croatia', 'Morocco ', 'Netherlands ',
       'England', 'Brazil', 'Portugal', 'Australia', 'USA', 'Poland ',
       'Senegal', 'Japan', 'South Korea', 'Spain', 'Switzerland',
       'Ecuador ', 'Qatar', 'Iran', 'Wales', 'Mexico ', 'Saudi Arabia',
       'Denmark', 'Tunisia', 'Germany', 'Costa Rica', 'Belgium',
       'Cameroon', 'Serbia', 'Ghana', 'Uruguay', 'Canada'], dtype=object)

We can clean these further by converting to title case and stripping strings of whitespace.

In [13395]:
wc_data_one['team'] = wc_data_one['team'].str.title().str.strip()

wc_data_two['team'] = wc_data_two['team'].str.title().str.strip()

Let's create a new DataFrame for analysis of goals by country using pd.DataFrame.

In [13396]:
highest_scoring_country = pd.DataFrame(columns=['Country', 'Goals'])

for country in countries_two:
    country_goals = wc_data_two.loc[wc_data_two['team'].str.lower() == country.lower(), 'goals'].sum()
    highest_scoring_country = pd.concat([highest_scoring_country, pd.DataFrame({'Country': [country], 'Goals': [country_goals]})], ignore_index=True)
        
print(highest_scoring_country)

         Country Goals
0      argentina    15
1         france    17
2        croatia     8
3        morocco     6
4    netherlands    10
5        england    13
6         brazil     8
7       portugal    12
8      australia     3
9            usa     3
10        poland     3
11       senegal     5
12         japan     5
13   south korea     3
14         spain    10
15   switzerland     5
16       ecuador     4
17         qatar     0
18          iran     4
19         wales     1
20        mexico     2
21  saudi arabia     3
22       denmark     1
23       tunisia     1
24       germany     6
25    costa rica     2
26       belgium     1
27      cameroon     4
28        serbia     5
29         ghana     5
30       uruguay     0
31        canada     1


# <span style="color: blue;">Exploratory Data Analysis</span>

### <span style="color: blue;">#1 - Who was the highest goalscorer?</span>

To determine the highest goalscorer, we can explore the **player_stats.csv** (wc_data_one) dataset.

In [13397]:
wc_data_one.sort_values(by='goals', ascending=False).head()

player       team                  age             club  minutes  \
377   Kylian Mbappé     France  23 years,  363 days         Paris SG      598   
388    Lionel Messi  Argentina  35 years,  177 days         Paris SG      690   
330  Julián Álvarez  Argentina  22 years,  321 days  Manchester City      464   
502  Olivier Giroud     France   36 years,  79 days            Milan      419   
672   Álvaro Morata      Spain   30 years,  56 days  Atlético Madrid      185   

     goals  assists  red cards  goals+assists  goals+assists per 90  \
377      8        2          0             10                 1.505   
388      7        3          0             10                 1.304   
330      4        0          0              4                 0.776   
502      4        0          0              4                 0.859   
672      3        1          0              4                 1.946   

     age in days  red cards per 90  
377         8758               0.0  
388        12952               0.0  
330         8351               0.0  
502        13219               0.0  
672        11006               0.0

Kylian Mbappé ranked the highest for goals scored, and even scored a hat trick in the final. This is a wild achievement (just one other player has achieved this - Geoff Hurst of England, against Germany 1966). However, his total (8) still doesn't come close to his late fellow Frenchman, Just Fontaine, who netted 13 times in 6 appearances during the 1958 competition!

__Click here to see his goals:__
<br>https://www.youtube.com/watch?v=5pzgfGgI4X8

### <span style="color: blue;">#2 - Who had the most assists?</span>

For the best playmaker, we can explore the **player_stats.csv** dataset, employing the top_assists_per_90 created during the Data Cleaning and Preparation stage.

In [13398]:
top_assists_per_90.sort_values(by='assists per 90', ascending=False)

player       team                  age             club  \
103    Bruno Fernandes   Portugal  28 years,  101 days   Manchester Utd   
240         Harry Kane    England  29 years,  143 days        Tottenham   
63   Antoine Griezmann     France  31 years,  272 days  Atlético Madrid   
273       Ivan Perišić    Croatia  33 years,  319 days        Tottenham   
388       Lionel Messi  Argentina  35 years,  177 days         Paris SG   

     minutes  goals  assists  red cards  assists per 90  
103      356      2        3          0           0.758  
240      402      2        3          0           0.672  
63       537      0        3          0           0.503  
273      669      1        3          0           0.404  
388      690      7        3          0           0.391

Bruno Fernandes with 0.758 assists per 90 minutes was the top playmaker of the competition - well done Bruno!

__Click here to check out his highlights:__
<br>https://www.youtube.com/watch?v=IkzTJWLYs64

### <span style="color: blue;">#3 - Who had the most goals and assists?</span>

Using the **player_stats.csv** (wc_data_one) dataset and the 'goals+assists' column created, we can determine the player with the most goals and assists.

In [13399]:
wc_data_one.sort_values(by='goals+assists', ascending=False).head()

player       team                  age             club  \
388     Lionel Messi  Argentina  35 years,  177 days         Paris SG   
377    Kylian Mbappé     France  23 years,  363 days         Paris SG   
103  Bruno Fernandes   Portugal  28 years,  101 days   Manchester Utd   
240       Harry Kane    England  29 years,  143 days        Tottenham   
330   Julián Álvarez  Argentina  22 years,  321 days  Manchester City   

     minutes  goals  assists  red cards  goals+assists  goals+assists per 90  \
388      690      7        3          0             10                 1.304   
377      598      8        2          0             10                 1.505   
103      356      2        3          0              5                 1.264   
240      402      2        3          0              5                 1.119   
330      464      4        0          0              4                 0.776   

     age in days  red cards per 90  
388        12952               0.0  
377         8758               0.0  
103        10321               0.0  
240        10728               0.0  
330         8351               0.0

Here, we have another tie between Messi and Mbappé - one could argue that Mbappé leads as he played less minutes, but I'll leave the winner of this one up to debate.

__Click here to watch the highlights from the final:__
<br>https://www.youtube.com/watch?v=zhEWqfP6V_w

### <span style="color: blue;">#4 - Who had the most goals and assists per 90 minutes?</span>


For this, we can use the **player_stats.csv** (wc_data_one) dataset and the 'goals+assists per 90' column created during preparation.

In [13400]:
wc_data_one.sort_values(by='goals+assists per 90', ascending=False).head()

player          team                  age           club  \
246   Hattan Bahebri  Saudi Arabia  30 years,  155 days       AlShabab   
483  Niclas Füllkrug       Germany  29 years,  312 days  Werder Bremen   
509    Paik Seung-ho   South Korea  25 years,  276 days        Jeonbuk   
649    Wout Weghorst   Netherlands  30 years,  133 days       Beşiktaş   
229    Gonçalo Ramos      Portugal  21 years,  181 days        Benfica   

     minutes  goals  assists  red cards  goals+assists  goals+assists per 90  \
246        4      0        1          0              1                22.500   
483       69      2        1          0              3                 3.913   
509       26      1        0          0              1                 3.462   
649       64      2        0          0              2                 2.812   
229      152      3        1          0              4                 2.368   

     age in days  red cards per 90  
246        11105               0.0  
483        10897               0.0  
509         9401               0.0  
649        11083               0.0  
229         7846               0.0

Hattan Bahebri from Saudi Arabia was the player with the most goals+assists per 90 minutes - 22.5!!! Wow.
<br>Although he only bagged one assist the whole competition, this is still a stat worth boasting about. He played just 4 minutes against Mexico, but made an instant impact, setting up Al-Dawsari inside the penalty area to get one back for his country.

__Skip to 1:35 to check it out:__
<br>https://www.youtube.com/watch?v=n55qabXypdQ

### <span style="color: blue;">#5 - Who had the most minutes played?</span>

Using **player_stats.csv** we can see which players had the most minutes.

In [13401]:
wc_data_one.sort_values(by='minutes', ascending=False).head(6)

player       team                  age           club  \
184  Emiliano Martínez  Argentina  30 years,  107 days    Aston Villa   
170  Dominik Livaković    Croatia  27 years,  343 days  Dinamo Zagreb   
491   Nicolás Otamendi  Argentina  34 years,  309 days        Benfica   
388       Lionel Messi  Argentina  35 years,  177 days       Paris SG   
325     Joško Gvardiol    Croatia  20 years,  329 days     RB Leipzig   
273       Ivan Perišić    Croatia  33 years,  319 days      Tottenham   

     minutes  goals  assists  red cards  goals+assists  goals+assists per 90  \
184      690      0        0          0              0                 0.000   
170      690      0        0          0              0                 0.000   
491      690      0        1          0              1                 0.130   
388      690      7        3          0             10                 1.304   
325      690      1        0          0              1                 0.130   
273      669      1        3          0              4                 0.538   

     age in days  red cards per 90  
184        11057               0.0  
170        10198               0.0  
491        12719               0.0  
388        12952               0.0  
325         7629               0.0  
273        12364               0.0

Interestingly, we have a tie between five players here. We will have to do a bit of research to make sure this information is correct.

After looking at a few different sources, I verified that this data was accurate. So, another tie. Arguably Martínez and Messi take this one as they won the competition (despite some questionable celebrations from Emiliano).

__Sources:__
<br>https://fbref.com/en/comps/1/stats/World-Cup-Stats
<br>https://www.statista.com/statistics/1356468/world-cup-qatar-most-minutes-played/


### <span style="color: blue;">#6 - Who were the youngest and oldest players?</span>

With **player_stats.csv**, and the 'age in days' column created, we can view the DataFrame in order of the youngest and oldest players.

In [13402]:
wc_data_one.sort_values(by='age in days', ascending=True).head()

player        team                  age           club  \
666  Youssoufa Moukoko     Germany   18 years,  28 days       Dortmund   
215        Garang Kuol   Australia   18 years,  94 days  Central Coast   
217               Gavi       Spain  18 years,  135 days      Barcelona   
294   Jewison Bennette  Costa Rica  18 years,  186 days     Sunderland   
94   Bilal El Khannous     Morocco  18 years,  222 days           Genk   

     minutes  goals  assists  red cards  goals+assists  goals+assists per 90  \
666        1      0        0          0              0                 0.000   
215       37      0        0          0              0                 0.000   
217      284      1        0          0              1                 0.317   
294      103      0        0          0              0                 0.000   
94        55      0        0          0              0                 0.000   

     age in days  red cards per 90  
666         6598               0.0  
215         6664               0.0  
217         6705               0.0  
294         6756               0.0  
94          6792               0.0

Youssoufa Moukoko from Germany was the youngest at 18 years and 28 days.

Northern Ireland's Norman Whiteside takes the record for the youngest player in World Cup history, at 17 years and 40 days. During his debut against Yugoslavia during the 1982 competition, he also became the youngest player to receive a yellow card in the competition to date.

In [13403]:
wc_data_one.sort_values(by='age in days', ascending=False).head()

player      team                  age            club  \
75    Atiba Hutchinson    Canada  39 years,  313 days        Beşiktaş   
519               Pepe  Portugal  39 years,  295 days           Porto   
142         Dani Alves    Brazil  39 years,  226 days            UNAM   
616       Thiago Silva    Brazil   38 years,  87 days         Chelsea   
134  Cristiano Ronaldo  Portugal  37 years,  316 days  Manchester Utd   

     minutes  goals  assists  red cards  goals+assists  goals+assists per 90  \
75       159      0        0          0              0                 0.000   
519      360      1        0          0              1                 0.250   
142      118      0        0          0              0                 0.000   
616      390      0        1          0              1                 0.231   
134      290      1        0          0              1                 0.310   

     age in days  red cards per 90  
75         14548               0.0  
519        14530               0.0  
142        14461               0.0  
616        13957               0.0  
134        13821               0.0

Atiba Hutchinson from Canada was the oldest player in this campaign, at 39 years and 313 days. 

Egyptian goalkeeper Essam El Hadary shadows this record as he took to the field during his team's final group match against Saudi Arabia at the 2018 World Cup, at 45 years and 161 days old, even saving a penalty.

### <span style="color: blue;">#7 - Who were the youngest and oldest players to score?</span>

With **player_stats.csv**, and the 'age in days' column created, we can view the DataFrame in order of the youngest and oldest players, looking for the first instance where the goals column value is more than zero.

I wasn't sure what the frame of reference was for the ages in this dataset, so I have included the ages at the time of the goal scored in the paragraphs below.

In [13404]:
# sorted the DataFrame age, finding a player with goals > 0
wc_data_one.sort_values(by=['age in days'], ascending=[True]).head()

player        team                  age           club  \
666  Youssoufa Moukoko     Germany   18 years,  28 days       Dortmund   
215        Garang Kuol   Australia   18 years,  94 days  Central Coast   
217               Gavi       Spain  18 years,  135 days      Barcelona   
294   Jewison Bennette  Costa Rica  18 years,  186 days     Sunderland   
94   Bilal El Khannous     Morocco  18 years,  222 days           Genk   

     minutes  goals  assists  red cards  goals+assists  goals+assists per 90  \
666        1      0        0          0              0                 0.000   
215       37      0        0          0              0                 0.000   
217      284      1        0          0              1                 0.317   
294      103      0        0          0              0                 0.000   
94        55      0        0          0              0                 0.000   

     age in days  red cards per 90  
666         6598               0.0  
215         6664               0.0  
217         6705               0.0  
294         6756               0.0  
94          6792               0.0

Gavi was the youngest to score with a tidy finish at the edge of the penalty area against Costa Rica, making him the third youngest player to score in a World Cup, being just 18 years and 110 days at the time. Legendary Brazilian attacker Pelé still holds the record for the youngest player ever to score in the World Cup, netting against Wales during the 1958 competition in Sweden.

__Check out the Gavi's goal here:__
<br>https://www.youtube.com/watch?v=JJbLXj5BkdU

In [13405]:
wc_data_one.sort_values(by=['age in days'], ascending=[False]).head()

player      team                  age            club  \
75    Atiba Hutchinson    Canada  39 years,  313 days        Beşiktaş   
519               Pepe  Portugal  39 years,  295 days           Porto   
142         Dani Alves    Brazil  39 years,  226 days            UNAM   
616       Thiago Silva    Brazil   38 years,  87 days         Chelsea   
134  Cristiano Ronaldo  Portugal  37 years,  316 days  Manchester Utd   

     minutes  goals  assists  red cards  goals+assists  goals+assists per 90  \
75       159      0        0          0              0                 0.000   
519      360      1        0          0              1                 0.250   
142      118      0        0          0              0                 0.000   
616      390      0        1          0              1                 0.231   
134      290      1        0          0              1                 0.310   

     age in days  red cards per 90  
75         14548               0.0  
519        14530               0.0  
142        14461               0.0  
616        13957               0.0  
134        13821               0.0

Pepe, the third oldest player in the tournament, bagged a goal against Switzerland, being 39 years and 283 days at the time. This made him the second oldest goalscorer at a World Cup! Not bad, but he still doesn't come close to the Cameroonian Roger Milla who scored against Russia at the USA 1994 edition at 42 years and 39 days. 
<br><br>Got one more in you Pepe?

### <span style="color: blue;">#8 - Who were the most carded players?</span>

Using the **player_stats.csv** dataset, and the 'red cards per 90' column created during preparation, we can see which players received red cards at the highest rate.

In [13406]:
wc_data_one.sort_values(by='red cards per 90', ascending=False).head()

player          team                  age             club  \
641     Walid Cheddira       Morocco  24 years,  330 days             Bari   
634  Vincent Aboubakar      Cameroon  30 years,  330 days          AlNassr   
644    Wayne Hennessey         Wales  35 years,  328 days  Nott'ham Forest   
163    Denzel Dumfries   Netherlands  26 years,  244 days            Inter   
458  Mohammed Al-Owais  Saudi Arabia   31 years,  69 days          AlHilal   

     minutes  goals  assists  red cards  goals+assists  goals+assists per 90  \
641       64      0        0          1              0                 0.000   
634      142      2        1          1              3                 1.901   
644      175      0        0          1              0                 0.000   
163      479      1        2          1              3                 0.564   
458      270      0        0          0              0                 0.000   

     age in days  red cards per 90  
641         9090             1.406  
634        11280             0.634  
644        13103             0.514  
163         9734             0.188  
458        11384             0.000

Walid Cheddira had the highest rate of red cards received, with 1.406 per 90 minutes.

### <span style="color: blue;">#9 - What country scored the most goals?</span>

Finally, using the country_goals DataFrame created during preparation from the **fifa_wc_2022_player_stats.csv** dataset, we can determine which country scored the most goals.

In [13407]:
highest_scoring_country.sort_values(by='Goals', ascending=False).head()

Country Goals
1       france    17
0    argentina    15
5      england    13
7     portugal    12
4  netherlands    10

We can see from this that France recorded the most goals in the competition, with 17. 
<br>Bien joué!

__Check out their highlights here:__
<br>https://www.youtube.com/watch?v=8Rr3eKPeJds

# <span style="color: red;">Conclusions</span>


From these datasets we were able to find out the following information:
- The highest scorer was Kylian Mbappé (8 goals)
- The best playmaker was Bruno Fernandes (0.758 assists per 90 minutes)
- The player with the most goals+assist was tied between Lionel Messi and Kylian Mbappé (10 goals+assists)
- The player with the most goals+assists per 90 minutes was Hattan Bahebri (22.5 goals+assists per 90 minutes)
- The player with the most minutes played was tied between Emiliano Martínez, Dominik Livaković, Nicolás Otamendi, Lionel Messi and Joško Gvardiol (690 minutes)
- The youngest+oldest players were Youssoufa Moukoko and Atiba Hutchinson, respectively (18 years and 28 days, 39 years and 313 days respectively)
- The youngest+oldest players to score were Gavi and Pepe (18 years and 110 days, 39 years and 283 days, respectively)
- The most red carded player was Walid Cheddira (1.406 red cards per 90 minutes)
- France was the highest scoring country (17 goals)

There are also some other ideas that are intriguing to explore, such as the highest scoring club, using match data to analyse individual games or seeing who the 'big game players' are i.e. who scores in important group stage games, finals, semi-finals etc.